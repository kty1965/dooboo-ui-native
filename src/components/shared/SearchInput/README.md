# Search Input

[![Npm Version](http://img.shields.io/npm/v/@dooboo-ui/native-search-input.svg?style=flat-square)](https://npmjs.org/package/@dooboo-ui/native-search-input)
[![Downloads](http://img.shields.io/npm/dm/@dooboo-ui/native-search-input.svg?style=flat-square)](https://npmjs.org/package/@dooboo-ui/native-search-input)

> Simple search input for react-native.

## Installation

```sh
yarn add @dooboo-ui/native
```

or

```sh
yarn add @dooboo-ui/native-search-input
```

## Usage

![](https://i.imgur.com/t84mZ1e.gif)

- debounced input
- reset button
- customizable icon

# Props

```ts
export interface SearchInputProps {
  value: string;
  onDebounceOrOnReset?: (value: string) => void;
  style?: StyleProp<ViewStyle>;
  debounceDelay?: number;
  customIcon?: React.ReactNode;
  placeholderText?: string;
}
```

|                   | necessary | types                  | default                  | info            |
| ----------------- | :-------: | ---------------------- | ------------------------ | --------------- |
| value             |     ✓     | string                 | undefined                |                 |
| onDebounceOrReset |     ✓     | (string) => void       | undefined                |                 |
| style             |           | `StyleProp<ViewStyle>` | undefined                | container style |
| debounceDelay     |           | number                 | 400                      |                 |
| customIcon        |           | React.ReactNode        | magnifier icon component |                 |
| placeholderText   |           | string                 | 검색어를 입력해주세요.   |                 |

# Getting started

- Import

  ```tsx
  import { SearchInput } from '@dooboo-ui/native';
  // or
  import SearchInput from '@dooboo-ui/native-search-input';
  ```

- Usage
  ```tsx
  const SearchInputWithState = () => {
    const [value, setValue] = useState('');
    return (
      <>
        <SearchInput
          value={value}
          onDebounceOrOnReset={setValue}
          debounceDelay={number('delay', 400)}
          placeholderText={text('placeholder', '')}
        />
        <Value>{`value (after debounced delay) : ${value}`}</Value>
      </>
    );
  };
  ```
