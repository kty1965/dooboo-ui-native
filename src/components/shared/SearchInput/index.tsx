import * as React from 'react';

import { StyleProp, View, ViewStyle } from 'react-native';

import styled from 'styled-components/native';

const Container = styled.View`
  background-color: #f3f5f7;
  height: 42;
  margin-left: 20;
  margin-right: 20;
  border-radius: 8;

  flex-direction: row;
`;

const Input = styled.TextInput`
  flex-grow: 1;
  align-self: center;
  font-size: 16;
`;

const ResetContainer = styled.View`
  height: 24;
  width: 24;
  margin-left: 2;
  margin-right: 11;
  margin-top: 9;
  margin-bottom: 9;
  justify-content: center;
  align-items: center;
`;

const Reset = styled.TouchableOpacity`
  background-color: #c6ccd1;
  border-radius: 20;
  width: 20;
  height: 20;
  justify-content: center;
  align-items: center;
`;

const ResetText = styled.Text`
  color: white;
`;

export interface SearchInputProps {
  testID?: string;
  value: string;
  onDebounceOrOnReset?: (value: string) => void;
  style?: StyleProp<ViewStyle>;
  debounceDelay?: number;
  customIcon?: React.ReactNode;
  placeholderText?: string;
}

// reference : https://dev.to/gabe_ragland/debouncing-with-react-hooks-jci
function useDebounce(value: string, delay = 400): string {
  const [debouncedValue, setDebouncedValue] = React.useState<string>(value);

  React.useEffect(() => {
    const handler = setTimeout(() => {
      setDebouncedValue(value);
    }, delay);
    return (): void => {
      clearTimeout(handler);
    };
  }, [value]);

  return debouncedValue;
}

const SearchInput: React.FC<SearchInputProps> = (props) => {
  const [value, setValue] = React.useState<string>(props.value);
  const debouncedValue = useDebounce(value, props.debounceDelay);

  React.useEffect(() => {
    if (props.onDebounceOrOnReset) {
      props.onDebounceOrOnReset(debouncedValue);
    }
  }, [debouncedValue]);

  React.useEffect(() => {
    setValue(props.value);
  }, [props.value]);

  return (
    <Container style={props.style}>
      {props.customIcon ? props.customIcon : <View style={{ width: 10 }} />}
      <Input
        testID={'SEARCH_INPUT'}
        value={value}
        onChangeText={(text): void => {
          setValue(text);
        }}
        placeholder={props.placeholderText || 'placehoder...'}
        placeholderTextColor={'#cdd2d7'}
      />
      {props.value !== '' && (
        <ResetContainer>
          <Reset
            testID={props.testID}
            onPress={(): void => {
              if (props.onDebounceOrOnReset) {
                props.onDebounceOrOnReset('');
              }
            }}
          >
            <ResetText>X</ResetText>
          </Reset>
        </ResetContainer>
      )}
    </Container>
  );
};

export default SearchInput;
