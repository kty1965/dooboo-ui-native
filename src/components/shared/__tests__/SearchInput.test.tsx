import * as React from 'react';

import renderer, { act } from 'react-test-renderer';

import SearchInput from '../SearchInput';
import { render } from '@testing-library/react-native';

const createTestProps = ({ value = '' }: { value: string }): object => ({
  testID: 'RESET_BUTTON',
  onDebounceOrOnReset: jest.fn(),
  value: value,
});

const component = (props): React.ReactElement => <SearchInput {...props} />;

describe('[SearchInput] render', () => {
  it('component and snapshot matches', () => {
    const rendered: renderer.ReactTestRendererJSON = renderer
      .create(component(createTestProps({ value: '' })))
      .toJSON();
    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });
});

describe('[SearchInput] interaction', () => {
  it('should show reset button when input `value` exists', () => {
    const rendered = render(
      component(createTestProps({ value: 'initial value' })),
    );
    expect(rendered.getByTestId('RESET_BUTTON')).toBeTruthy();
  });

  it('should not show reset button when empty input value', () => {
    const rendered = render(component(createTestProps({ value: '' })));
    expect(rendered.queryByTestId('RESET_BUTTON')).toBeNull();
  });
});
