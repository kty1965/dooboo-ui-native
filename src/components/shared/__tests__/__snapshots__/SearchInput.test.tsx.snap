// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[SearchInput] render component and snapshot matches 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#f3f5f7",
        "borderRadius": 8,
        "flexDirection": "row",
        "height": 42,
        "marginLeft": 20,
        "marginRight": 20,
      },
    ]
  }
>
  <View
    style={
      Object {
        "width": 10,
      }
    }
  />
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="placehoder..."
    placeholderTextColor="#cdd2d7"
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "alignSelf": "center",
          "flexGrow": 1,
          "fontSize": 16,
        },
      ]
    }
    testID="SEARCH_INPUT"
    underlineColorAndroid="transparent"
    value=""
  />
</View>
`;
